File: leetcode2134/solution.cpp
/*
Solution 1:

Sliding Window

Time: O(n) | Space: O(1)

- n: length of 'nums'
*/

class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int n = nums.size();
        int ones = 0;

        for (int &num: nums) {
            ones += num;
        }

        int sum = 0;

        for (int i = 0; i < ones; ++i) {
            sum += nums[i];
        }

        int res = ones - sum;

        for (int i = 0; i < n; ++i) {
            sum += nums[(i + ones) % n] - nums[i];

            res = min(res, ones - sum);
        }

        return res;
    }
};

File: leetcode2134/solution.rs
/*
Solution 1:

Sliding Window

Time: O(n) | Space: O(1)

- n: length of nums
*/

impl Solution {
    pub fn min_swaps(nums: Vec<i32>) -> i32 {
        let n: usize = nums.len();

        let mut ones: i32 = 0;

        for &num in nums.iter() {
            ones += num;
        }

        let mut sum: i32 = 0;

        for i in 0..ones as usize {
            sum += nums[i];
        }

        let mut res: i32 = ones - sum;

        for i in 0..n {
            sum += nums[(i + ones as usize) % n] - nums[i];

            res = res.min(ones - sum);
        }

        res
    }
}